apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,322 @@
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified gRPC call invoker.
+        /// </summary>
+        /// <para>
+        /// This method removes any credentials specified with previous calls.
+        /// </para>
+        /// <param name="callInvoker">The gRPC call invoker to use, or null to use the default call invoker.</param>
+        /// <returns>A builder with the specified call invoker.</returns>
+        public TBuilder WithCallInvoker(CallInvoker callInvoker)
+        {
+            var clone = Clone();
+            clone.ClearCredentials();
+            clone.CallInvoker = callInvoker;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a call invoker synchronously. Override this method in a concrete builder type if more
+        /// call invoker mechanisms are supported.
+        /// </summary>
+        protected virtual CallInvoker CreateCallInvoker()
+        {
+            if (CallInvoker != null)
+            {
+                return CallInvoker;
+            }
+            var endpoint = Endpoint ?? GetDefaultEndpoint();
+            var credentials = GetChannelCredentials();
+            Channel channel = new Channel(endpoint.Host, endpoint.Port, GetChannelCredentials());
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Creates a call invoker asynchronously. Override this method in a concrete builder type if more
+        /// call invoker mechanisms are supported.
+        /// </summary>
+        protected virtual async Task<CallInvoker> CreateCallInvokerAsync(CancellationToken cancellationToken)
+        {
+            if (CallInvoker != null)
+            {
+                return CallInvoker;
+            }
+            var endpoint = Endpoint ?? GetDefaultEndpoint();
+            var credentials = await GetChannelCredentialsAsync(cancellationToken).ConfigureAwait(false);
+            Channel channel = new Channel(endpoint.Host, endpoint.Port, credentials);
+            return new DefaultCallInvoker(channel);
+        }
+
+        /// <summary>
+        /// Obtains channel credentials synchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected virtual ChannelCredentials GetChannelCredentials()
+        {
+            if (ChannelCredentials != null)
+            {
+                return ChannelCredentials;
+            }
+            GoogleCredential unscoped;
+            if (CredentialsFile != null)
+            {
+                unscoped = GoogleCredential.FromFile(CredentialsFile);
+            }
+            else if (JsonCredentials != null)
+            {
+                unscoped = GoogleCredential.FromJson(JsonCredentials);
+            }
+            else
+            {
+                unscoped = GoogleCredential.GetApplicationDefault();
```

Comment:
I would put this in a virtual GetDefaultCredentials() method and call here the GetDefaultCredentials() method in its place.

Reply:
Any particular reason? I could see it perhaps for testing, but that's all.